src/Ledger/Ledger/Properties.agda

```diff
@@ -82,29 +102,579 @@ instance
+      (EPOCH x) govDepsMatch-ls =
+      ≡ᵉ.trans (filter-pres-≡ᵉ $ dom-cong (res-comp-cong (≡ᵉ.sym χ'≡χ))) (main-invariance-lemma govDepsMatch-ls)
+      where
+
+      -- the combinator used in the EPOCH rule
+      χ : (DepositPurpose ⇀ Coin) → ℙ DepositPurpose
+      χ deps = mapˢ (proj₁ ∘ proj₂)
+        (flip concatMapˢ removed  (λ (gaid , gast) → mapˢ (returnAddr gast ,_) ((deps ∣ ❴ GovActionDeposit gaid ❵) ˢ)))
+
+      -- a simpler combinator that suffices here;
+      χ' : ℙ DepositPurpose
+      χ' = mapˢ (GovActionDeposit ∘ proj₁) removed
+      -- Below we prove χ and χ' are essentially equivalent.
+
+      P : GovActionID × GovActionState → Set
+      P = λ u → proj₁ u ∉ mapˢ proj₁ removed
+
+      Pᵇ : GovActionID × GovActionState → Bool
+      Pᵇ = λ u → does ¿ P u ¿
+
+      utxoDeps : Deposits
+      utxoDeps = UTxOState.deposits utxoSt
+
+      -- utxo deposits restricted to new form of set used in EPOCH rule
+      utxoDeps' : Deposits
+      utxoDeps' = utxoDeps ∣ χ' ᶜ
+
+      -- utxo deposits restricted to old form of set used in EPOCH rule
+      utxoDeps'' : Deposits
+      utxoDeps'' = utxoDeps ∣ (χ utxoDeps) ᶜ
+
+      open Equivalence
+
+      χ'≡χ : χ' ≡ᵉ χ utxoDeps
+      χ'≡χ = χ'⊆χ , χ⊆χ'
+        where
+        χ'⊆χ : χ' ⊆ χ utxoDeps
+        χ'⊆χ {a} x with from ∈-map x
+        ... | ((gaid , gast) , a≡GAgaid , gaidgast∈rem) with from ∈-map (ratify-removed x)
+        ... | ((dp , c) , a≡dp , dpc∈utxoDeps) =
+          to ∈-map ( (returnAddr gast , GovActionDeposit gaid , c)
+                   , a≡GAgaid
+                   , to ∈-concatMapˢ((gaid , gast) , gaidgast∈rem , addr-dp-c-∈map)
+                   )
+          where
+          dp-c-∈-utxoDep : (GovActionDeposit gaid , c) ∈  (utxoDeps ˢ)
+          dp-c-∈-utxoDep = subst (λ x₂ → (x₂ , c) ∈ (utxoDeps ˢ))
+                                 (trans (sym a≡dp) a≡GAgaid) dpc∈utxoDeps
+
+          addr-dp-c-∈map : (returnAddr gast , GovActionDeposit gaid , c)
+                           ∈ mapˢ (returnAddr gast ,_) ( (utxoDeps ∣ ❴ GovActionDeposit gaid ❵)ˢ )
+          addr-dp-c-∈map =
+            to ∈-map $ (GovActionDeposit gaid , c)
+                     , (×-≡,≡→≡ (refl , refl) , res-singleton⁺{m = utxoDeps} dp-c-∈-utxoDep)
+        χ⊆χ' : χ utxoDeps ⊆ χ'
+        χ⊆χ' {a} x with (from ∈-map x)
+        ... | ((rwa , dp , c) , a≡dp , rwa-dp-c∈) with (from ∈-concatMapˢ rwa-dp-c∈)
+        ... | ((gaid , gast) , gaid-gast-∈-removed , rwa-dp-c-∈-map) with (from ∈-map rwa-dp-c-∈-map)
+        ... | ((dp' , _) , q≡ , q∈) = to ∈-map ((gaid , gast) , a≡GA , gaid-gast-∈-removed)
+          where
+          a≡GA : a ≡ GovActionDeposit gaid
+          a≡GA = let open ≡-Reasoning in begin
+            a                       ≡⟨ a≡dp ⟩
+            dp                      ≡⟨ cong (proj₁ ∘ proj₂) q≡ ⟩
+            dp'                     ≡⟨ cong proj₁ $ proj₂ (res-singleton''{m = utxoDeps} $ q∈) ⟩
+            GovActionDeposit gaid   ∎
```

Comment:
Can you not do this proof without splitting up the two inclusions? I see a lot of things like `from ∈-map x`, which reads to me like what you should actually be doing would be to use `∈-map` to rewrite your expression. It would be something like this:
```agda
... = from ≡ᵉ⇔≡ᵉ' λ a → let open R.EquationalReasoning in
        ...
          ∼⟨ ∈-map ⟩
        ...
```
And even if you cannot fully complete the proof this way, you can still eliminate a bunch of the repetition.

Reply:
This was one of the harder, more technical proofs in the PR.  It is quite delicate/brittle and already required a substantial amount of time/effort to get it right and then simplified it to where it is now.  I can try to make some minor improvements, but doing both inclusions simultaneously, while maybe possible, will be challenging for me and I'm not sure it's worth the additional time/effort.

Reply:
I simplified the proof as much as I could... still can't see how to make it bidirectional.